### ISSUE TITLE
[Python SDK] slow performance with large CRD in helm chart.

### ISSUE THREAD

**Issue by misterkramer** [pulumi/pulumi#8717]
## Expected Behavior

`pulumi up` is fast just like in typescript ~5 seconds
[Typescript SDK TRACE out.txt](https://github.com/pulumi/pulumi/files/5795889/out.txt)

## Current Behavior

`pulumi up` with python is very very slow ~1 hour !
[Python SDK TRACE out.txt.zip](https://github.com/pulumi/pulumi/files/5796026/out.txt.zip)
(I didn't wait and canceled the command `pulumi up`, very long wait)
## Steps to Reproduce

1) Chart with a large CRD yaml size ~10k lines
Example:
https://github.com/VictoriaMetrics/helm-charts/blob/master/charts/victoria-metrics-operator/templates/crd.yaml
https://github.com/VictoriaMetrics/helm-charts/tree/master/charts/victoria-metrics-operator

2) use python sdk to install chart:
```
from pulumi_kubernetes.helm import v3
from pulumi_kubernetes.helm.v3.helm import ChartOpts


vm_chart = v3.Chart("vm", ChartOpts(chart="victoria-metrics-operator", repo="vm"))
```

## Context (Environment)

```
macOS Big Sur 11.1
Python 3.9.1
pulumi v2.17.0
pulumi-kubernetes==2.7.4
```

**Comment by justinvp** [comment]
I haven't had a chance to try this, but we should be able to use [pyinstrument](https://github.com/joerick/pyinstrument) to profile the Python program:

```python
from pyinstrument import Profiler
profiler = Profiler()
profiler.start()

from pulumi_kubernetes.yaml import ConfigGroup

def remove_status_field(obj):
    if obj["kind"] == "CustomResourceDefinition" and 'status' in obj:
        del obj["status"]

ConfigGroup(
    "test",
    skip_await=False,
    files=["https://raw.githubusercontent.com/traefik/traefik/ba822acb23db3d620f6b3773ac9eff3e98ce313f/integration/fixtures/k8s/01-traefik-crd.yml"],
    transformations=[remove_status_field],
)

profiler.stop()
with open("profile.html", 'w') as f:
    f.write(profiler.output_html())
```

**Comment by lblackstone** [comment]
@justinvp I profiled like you suggested, and the Python program took < 3 seconds. Seems like the bulk of the time is waiting on the engine.
![Screen Shot 2022-01-11 at 4 39 55 PM](https://user-images.githubusercontent.com/9253463/149038400-52bf56b0-3091-4d53-a0e4-1a1e2799e8b6.png)

**Comment by iwahbe** [comment]
> @nathanielobrown Thanks for the information!
> 
> I was able to reproduce this slowness with the following program on my M1 Mac.
> 
> ```python
> from pulumi_kubernetes.yaml import ConfigGroup
> 
> def remove_status_field(obj):
>     if obj["kind"] == "CustomResourceDefinition" and 'status' in obj:
>         del obj["status"]
> 
> ConfigGroup(
>     "test",
>     skip_await=False,
>     files=["https://raw.githubusercontent.com/traefik/traefik/ba822acb23db3d620f6b3773ac9eff3e98ce313f/integration/fixtures/k8s/01-traefik-crd.yml"],
>     transformations=[remove_status_field],
> )
> ```

@lblackstone This completed on my (non M1) Mac in 19s against a machine local cluster. Do you think this relates to your laptop being M1? Or do you think it relates to the local/remote cluster?

**Comment by lblackstone** [comment]
> @lblackstone This completed on my (non M1) Mac in 19s against a machine local cluster. Do you think this relates to your laptop being M1? Or do you think it relates to the local/remote cluster?

I tested this on a local cluster as well, so I'd guess it has something to do with the M1 if you're not seeing the same slowdown.

**Comment by rawkode** [comment]
I've ran this on my M1 machine and I am seeing the same slowness.

Here's the about:

```shell
❯ pulumi about
CLI
Version      3.22.1
Go Version   go1.17.5
Go Compiler  gc

Host
OS       darwin
Version  12.1
Arch     arm64

Backend
Name     rawkode-mbp14.5gee.router
URL      file://~
User     rawkode
```

Here's the preview:

```shell
❯ time pulumi preview
Enter your passphrase to unlock config/secrets
    (set PULUMI_CONFIG_PASSPHRASE or PULUMI_CONFIG_PASSPHRASE_FILE to remember):
Previewing update (devpy):
     Type                                                                  Name                                                                                                                                    Plan
 +   pulumi:pulumi:Stack                                                   tmp.E8UNNkB8-devpy                                                                                                                      create
 +   └─ kubernetes:yaml:ConfigGroup                                        test                                                                                                                                    create
 +      └─ kubernetes:yaml:ConfigFile                                      https://raw.githubusercontent.com/traefik/traefik/ba822acb23db3d620f6b3773ac9eff3e98ce313f/integration/fixtures/k8s/01-traefik-crd.yml  create
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  tlsstores.traefik.containo.us                                                                                                           create
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  middlewaretcps.traefik.containo.us                                                                                                      create
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  tlsoptions.traefik.containo.us                                                                                                          create
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  serverstransports.traefik.containo.us                                                                                                   create
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  ingressrouteudps.traefik.containo.us                                                                                                    create
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  ingressroutetcps.traefik.containo.us                                                                                                    create
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  middlewares.traefik.containo.us                                                                                                         create
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  ingressroutes.traefik.containo.us                                                                                                       create
 +         └─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  traefikservices.traefik.containo.us                                                                                                     create

Resources:
    + 12 to create

pulumi preview  220.81s user 1.01s system 99% cpu 3:43.55 total
```

**Comment by stevesloka** [comment]
For another data point, I ran on my first-gen M1 and got just about the same results as @rawkode:

```bash
pulumi preview  225.07s user 1.05s system 98% cpu 3:48.43 total
```

**Comment by mikhailshilkov** [comment]
Another non-M1 data point:

```
> pulumi about
CLI          
Version      3.22.1
Go Version   go1.17.5
Go Compiler  gc

Plugins
NAME          VERSION
azure-native  1.54.0
azuread       4.3.0
kubernetes    3.15.0
python        unknown
random        4.3.1
tls           4.1.0

Host     
OS       darwin
Version  12.1
Arch     x86_64

> time pulumi preview
...
pulumi preview  22.04s user 2.64s system 62% cpu 39.210 total
```

So the problem does seem to be M1-specific.

**Comment by iwahbe** [comment]
Thanks for confirming @mikhailshilkov

**Comment by farvour** [comment]
> For another data point, I ran on my first-gen M1 and got just about the same results as @rawkode:
> 
> ```shell
> pulumi preview  225.07s user 1.05s system 98% cpu 3:48.43 total
> ```

Fascinating. I just ran into this as well. I'm using the x86 binary to keep compatibility with providers that are x86 only (no aarch64) and I bet this is something to do with Python's internals w/r/t Rosetta 2.

**Comment by cbcmg** [comment]
This could be helpful: https://github.com/pulumi/pulumi-kubernetes/issues/1867#issuecomment-1107353905

**Comment by mikhailshilkov** [comment]
@Zaid-Ajaj As you are the only person with an M1 on our team, I'm assigning this to you to take a look. Thank you!

**Comment by Zaid-Ajaj** [comment]
I have looked into this and can confirm the findings on an M1 that the python is really slow
```
(venv) zaid@Zaids-MBP slow-python % time pulumi preview
Previewing update (dev)

View Live: https://app.pulumi.com/Zaid-Ajaj/slow-python/dev/previews/7f75858f-0d0a-4c3b-a76d-a6231a32365b

     Type                                                                  Name                                                                                                       
 +   pulumi:pulumi:Stack                                                   slow-python-dev                                                                                            
 +   └─ kubernetes:yaml:ConfigGroup                                        test                                                                                                   
 +      └─ kubernetes:yaml:ConfigFile                                      https://raw.githubusercontent.com/traefik/traefik/ba822acb23db3d620f6b3773ac9eff3e98ce313f/integration/
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  tlsstores.traefik.containo.us                                                                          
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  middlewaretcps.traefik.containo.us                                                                     
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  tlsoptions.traefik.containo.us                                                                         
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  serverstransports.traefik.containo.us                                                                  
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  ingressrouteudps.traefik.containo.us                                                                   
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  ingressroutetcps.traefik.containo.us                                                                   
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  middlewares.traefik.containo.us                                                                        
 +         ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  ingressroutes.traefik.containo.us                                                                      
 +         └─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  traefikservices.traefik.containo.us                                                                    
 
Resources:
    + 12 to create

pulumi preview  220.55s user 1.32s system 99% cpu 3:43.74 total
``` 
This is happening because of https://github.com/protocolbuffers/protobuf/issues/9839 where the pure python implementation of protobuf is used on Apple Silicon instead of the C++ implementation. That said, the [Protocol Buffers v[3].21.0](https://github.com/protocolbuffers/protobuf/releases/tag/v21.0) release includes:
```
Python: 
- [...]
- The binary wheel for macOS now supports Apple silicon.
```
Which means that when we upgrade our python SDK to use this release, the SDK will no longer use the pure python implementation, but the C++ one instead. 

https://github.com/pulumi/pulumi/pull/9713 is the first step towards that upgrade

**Comment by dpunkturban** [comment]
I have the same problem on my M1 (Max). Pulumi is no longer usable :/ Is there already an ETA when the problem will be solved?

**Comment by Zaid-Ajaj** [comment]
@dpunkturban @misterkramer This issue will be resolved when #10063 is merged. I have confirmed that the fix works on my local M1 machine and that `pulumi preview` took 1.14s with the update as opposed to the 220.55s that I got before. 

We will merge the PR as soon as possible and the fix will become available in the next pulumi release, either this week or next week Wednesday
